official/mnist/mnist.py

```diff
@@ -75,141 +76,128 @@ def eval_dataset(data_dir):
+  if isinstance(image, dict):
+    image = image['image']
+  logits = model(image, training=False)
   predictions = {
-      'classes': tf.argmax(input=logits, axis=1),
-      'probabilities': tf.nn.softmax(logits, name='softmax_tensor')
+      'classes': tf.argmax(logits, axis=1),
+      'probabilities': tf.nn.softmax(logits),
   }
+  return tf.estimator.EstimatorSpec(
+      mode=tf.estimator.ModeKeys.PREDICT,
+      predictions=predictions,
+      export_outputs={
+          'classify': tf.estimator.export.PredictOutput(predictions)
+      })
 
-  if mode == tf.estimator.ModeKeys.PREDICT:
-    export_outputs={'classify': tf.estimator.export.PredictOutput(predictions)}
-    return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions,
-                                      export_outputs=export_outputs)
 
+def train_spec(model, image, labels):
+  """EstimatorSpec for training."""
+  optimizer = tf.train.AdamOptimizer(learning_rate=1e-4)
+  logits = model(image, training=True)
   loss = tf.losses.softmax_cross_entropy(onehot_labels=labels, logits=logits)
-
-  # Configure the training op
-  if mode == tf.estimator.ModeKeys.TRAIN:
-    optimizer = tf.train.AdamOptimizer(learning_rate=1e-4)
-    train_op = optimizer.minimize(loss, tf.train.get_or_create_global_step())
-  else:
-    train_op = None
-
   accuracy = tf.metrics.accuracy(
-      tf.argmax(labels, axis=1), predictions['classes'])
-  metrics = {'accuracy': accuracy}
-
-  # Create a tensor named train_accuracy for logging purposes
+      labels=tf.argmax(labels, axis=1), predictions=tf.argmax(logits, axis=1))
+  # Name the accuracy tensor 'train_accuracy' to demonstrate the
+  # LoggingTensorHook.
   tf.identity(accuracy[1], name='train_accuracy')
   tf.summary.scalar('train_accuracy', accuracy[1])
+  return tf.estimator.EstimatorSpec(
+      mode=tf.estimator.ModeKeys.TRAIN,
+      loss=loss,
+      train_op=optimizer.minimize(loss, tf.train.get_or_create_global_step()))
 
+
+def eval_spec(model, image, labels):
+  """EstimatorSpec for evaluation."""
+  logits = model(image, training=False)
+  loss = tf.losses.softmax_cross_entropy(onehot_labels=labels, logits=logits)
   return tf.estimator.EstimatorSpec(
-      mode=mode,
-      predictions=predictions,
+      mode=tf.estimator.ModeKeys.EVAL,
       loss=loss,
-      train_op=train_op,
-      eval_metric_ops=metrics)
+      eval_metric_ops={
+          'accuracy':
+              tf.metrics.accuracy(
+                  labels=tf.argmax(labels, axis=1),
+                  predictions=tf.argmax(logits, axis=1)),
+      })
+
+
+def model_fn(features, labels, mode, params):
+  """The model_fn argument for creating an Estimator."""
+  model = Model(params['data_format'])
+  if mode == tf.estimator.ModeKeys.PREDICT:
+    return predict_spec(model, features)
+  if mode == tf.estimator.ModeKeys.TRAIN:
+    return train_spec(model, features, labels)
+  if mode == tf.estimator.ModeKeys.EVAL:
+    return eval_spec(model, features, labels)
```

Comment:
It makes me a little bit sad that this is split up into three separate functions. Do you need the separate functions? 

Would it make sense to simply always fill out all fields? Or did you try to show what is needed in which case more explicitly?

Even if you want to show it explicitly, I would probably prefer to inline these, but I haven't seen the resulting function, maybe it is too big.

Reply:
As a newbie to using estimators, I found it helpful to understand what is needed when. So I would like to avoid filling out all the fields. That way one doesn't have to think about loss functions to get a predictor up and running fast and other such things.

Separate functions or inlining - my personal preference was to have them separate, but I don't have a strong argument either way. Will inline.

Reply:
Done
(Though, after doing it I feel slightly more strongly that the separate functions made it more readable, but I realize that I'm not the authority on how other people should read code :)

Reply:
Another alternative is to always fill in all fields. This is what we're recommending, and since this is graph mode, it is harmless (that may change if Estimators ever support Eager).

If you wanted to make clear what is needed in which mode, that's not a great option of course, but it makes it more concise, and the cost in the graph is trivial.

Reply:
I agree the cost in the graph is trivial.
However, I feel it's conceptually cleaner to not have to fill-in all fields. If I'm interested only in predictions, would be nice to not have to read through code about optimizers and losses. If I'm interested only in training, would be nice to not have to read through code around metrics.

Filling all fields won't make it much more concise either - the 2-3 duplicated lines are worth it, IMHO.

Let me know if you strongly disagree :)